storage/replica_command.go

```diff
@@ -614,6 +540,92 @@ func (r *Replica) EndTransaction(
 	return reply, externalIntents, nil
 }
 
+// resolveExplicitIntents resolve any explicit intents and persists e
+// transaction record with updated status. All that are local to this
+// range get resolved synchronously in the same batch. The remainder
+// are collected and returned so that they can be handed off to
+// asynchronous processing.
+func (r *Replica) resolveExplicitIntents(batch engine.Engine, ms *engine.MVCCStats, args roachpb.EndTransactionRequest, txn *roachpb.Transaction) ([]roachpb.Intent, error) {
+	desc := r.Desc()
+	var preMergeDesc *roachpb.RangeDescriptor
+	if mergeTrigger := args.InternalCommitTrigger.GetMergeTrigger(); mergeTrigger != nil {
+		// If this is a merge, then use the post-merge descriptor to determine
+		// which intents are local (note that for a split, we want to use the
+		// pre-split one instead because it's larger).
+		preMergeDesc = desc
+		desc = &mergeTrigger.UpdatedDesc
+	}
+
+	iterAndBuf := engine.GetIterAndBuf(batch)
+	defer iterAndBuf.Cleanup()
+
+	var externalIntents []roachpb.Intent
+	for _, span := range args.IntentSpans {
+		if err := func() error {
+			intent := roachpb.Intent{Span: span, Txn: txn.TxnMeta, Status: txn.Status}
+			if len(span.EndKey) == 0 {
+				// For single-key intents, do a KeyAddress-aware check of
+				// whether it's contained in our Range.
+				if !containsKey(*desc, span.Key) {
+					externalIntents = append(externalIntents, intent)
+					return nil
+				}
+				resolveMs := ms
+				if preMergeDesc != nil && !containsKey(*preMergeDesc, span.Key) {
+					// If this transaction included a merge and the intents
+					// are from the subsumed range, ignore the intent resolution
+					// stats, as they will already be accounted for during the
+					// merge trigger.
+					resolveMs = nil
+				}
+				return engine.MVCCResolveWriteIntentUsingIter(batch, iterAndBuf, resolveMs, intent)
+			}
+			// For intent ranges, cut into parts inside and outside our key
+			// range. Resolve locally inside, delegate the rest. In particular,
+			// an intent range for range-local data is correctly considered local.
+			inSpan, outSpans := intersectSpan(span, *desc)
+			for _, span := range outSpans {
+				outIntent := intent
+				outIntent.Span = span
+				externalIntents = append(externalIntents, outIntent)
+			}
+			if inSpan != nil {
+				intent.Span = *inSpan
+				_, err := engine.MVCCResolveWriteIntentRangeUsingIter(batch, iterAndBuf, ms, intent, 0)
+				return err
+			}
+			return nil
+		}(); err != nil {
+			// TODO(tschottdorf): any legitimate reason for this to happen?
+			// Figure that out and if not, should still be ReplicaCorruption
+			// and not a panic.
+			panic(fmt.Sprintf("error resolving intent at %s on end transaction [%s]: %s", span, txn.Status, err))
+		}
+	}
+
+	// Persist the transaction record with updated status (& possibly timestamp).
```

Comment:
It's weird that everything that happens from here on down happens inside of this helper method. Auto-gc'ing can easily move outside. Writing an updated txn entry can maybe stay, but the method name needs to indicate that this is happening (and nobody can write another entry in `EndTransaction` after this one unless it reads the entry from disk first).